src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs

```diff
@@ -0,0 +1,513 @@
+			{
+				1 << SCROLL_RIGHT => (-ScrollContentPresenter.ScrollViewerDefaultMouseWheelDelta, true),
+				1 << SCROLL_DOWN => (-ScrollContentPresenter.ScrollViewerDefaultMouseWheelDelta, false),
+				1 << SCROLL_LEFT => (ScrollContentPresenter.ScrollViewerDefaultMouseWheelDelta, true),
+				1 << SCROLL_UP => (ScrollContentPresenter.ScrollViewerDefaultMouseWheelDelta, false),
+				_ => (0, false)
+			};
+		}
+
+		var info = GetDevicePropertiesFromId(data.sourceid);
+
+		// Touchpad scrolling manifests in a Motion event where the current scrolling "position" is recorded in the
+		// valuator. There is no direct way to get the delta, so we have to record the old position and diff it
+		// with the new position. This also means that the first "tick" will not result in a WheelChanged event,
+		// but will only be used to set the initial wheel position.
+		// Also, we can get a "diagonal scroll" where both the horizontal and the vertical positions change.
+		// Our PointerEventArgs don't support this, so we arbitrarily choose to make diagonal scrolling
+		// result in a vertical scroll.
+		// IMPORTANT: DO NOT FORGET TO RESET POSITIONS ON LEAVE.
+		if (data.evtype is XiEventType.XI_Motion && info is { } info_)
+		{
+			var valuators = data.valuators;
+			var values = valuators.Values;
+			for (var i = 0; i < valuators.MaskLen * 8; i++)
+			{
+				if (XLib.XIMaskIsSet(valuators.Mask, i))
+				{
+					var (valuator, value) = (i, *values++);
+					if (valuator == info_.HorizontalValuator || valuator == info_.VerticalValuator)
+					{
+						isHorizontalMouseWheel = valuator == info_.HorizontalValuator;
+						var (dict, increment) = isHorizontalMouseWheel ?
+							(_lastHorizontalTouchpadWheelPosition, info_.HorizontalIncrement!.Value) :
+							(_lastVerticalTouchpadWheelPosition, info_.VerticalIncrement!.Value);
+						if (dict.TryGetValue(data.sourceid, out var oldValue))
+						{
+							wheelDelta = (oldValue - value) * ScrollContentPresenter.ScrollViewerDefaultMouseWheelDelta / increment;
+						}
+						dict[data.sourceid] = value;
+					}
+					// DO NOT BREAK OUT OF THE FOR LOOP. We still need to update the other valuator positions.
+				}
+			}
+		}
+
+		var properties = new PointerPointProperties
+		{
+			IsLeftButtonPressed = (mask & (1 << LEFT)) != 0,
+			IsMiddleButtonPressed = (mask & (1 << MIDDLE)) != 0,
+			IsRightButtonPressed = (mask & (1 << RIGHT)) != 0,
+			IsXButton1Pressed = (mask & (1 << XButton1)) != 0,
+			IsXButton2Pressed = (mask & (1 << XButton2)) != 0,
+			IsHorizontalMouseWheel = isHorizontalMouseWheel,
+			IsInRange = true,
+			IsTouchPad = info is { } && IsTouchpad(info.Value.Properties),
+			MouseWheelDelta = (int)Math.Round(wheelDelta)
+			// TODO: PointerUpdateKind is currently not used for anything, so we can leave it blank
+			// for now, but ideally we should probably be diffing the previous and new button states
+			// and computing an update kind from that.
+		};
```

Comment:
How about add the touch Pressure and touch width and height here? Thank you.

Reply:
We can do MT protocol (https://www.kernel.org/doc/html/latest/input/multi-touch-protocol.html)-related stuff in a separate PR if there's enough demand.